torch/storage.py

```diff
@@ -237,6 +284,18 @@ def __getitem__(self, *args, **kwargs):
     def is_cuda(self):
         return self.device.type == 'cuda'
 
+    @_share_memory_lock_protected
+    def share_memory_(self, *args, **kwargs):
```

Comment:
The problem is that our serialization code calls into the internal functions directly. So I have to protect all of them.
I moved the locks to be reentrant locks to avoid deadlock when share_memory_ calls back into _share_fd